### PR TITLE
BUG: don't sort unique values from categoricals

### PR DESCRIPTION
**Pull request by shoyer**
This should resolve the inconsistency @mwaskom reported in #9148.

CC @jreback @TomAugspurger @JanSchulz

### CODE REVIEW

**Comment by jreback** [comment]
this makes sense

**Comment by mwaskom** [comment]
:+1: thanks @shoyer

File: pandas/tests/test_categorical.py [hunk 1]
```diff
@@ -774,14 +774,27 @@ def test_unique(self):
...
         cat = Categorical(["a","b","a","a"], categories=["a","b","c"])
         res = cat.unique()
         self.assert_numpy_array_equal(res, exp)
+
         cat = Categorical(["a","b","a", np.nan], categories=["a","b","c"])
         res = cat.unique()
         exp = np.asarray(["a","b", np.nan], dtype=object)
         self.assert_numpy_array_equal(res, exp)
 
+        cat = Categorical(["b", "a"], categories=["a", "b"], ordered=False)
```
Comment (JanSchulz) [inline pandas/tests/test_categorical.py]: A comment would be nice for the next coder :-)

**Comment by JanSchulz** [comment]
LGTM

File: pandas/core/categorical.py [hunk 2]
```diff
@@ -1388,15 +1388,22 @@ def unique(self):
...
         -------
         unique values : array
         """
-        unique_codes = np.unique(self.codes)
-        # for compatibility with normal unique, which has nan last
-        if unique_codes[0] == -1:
-            unique_codes[0:-1] = unique_codes[1:]
-            unique_codes[-1] = -1
+        from pandas.core.nanops import unique1d
+        unique_codes = unique1d(self.codes)
```
Comment (JanSchulz) [inline pandas/core/categorical.py]: A comment on the difference between `np.unique` and `unique1d` and why the latter is used would be nice.

**Comment by shoyer** [comment]
Actually, does it even make sense to sort the unique values for even ordered Categoricals? `.unique()` does not sort for any other dtype, and it's easy enough to chain `.unique().order()`.

**Comment by jorisvandenbossche** [comment]
@shoyer I just wanted to ask a similar question: what if you want the unique values in order of appearance when you have an ordered categorical? That is now rather impractical to obtain.

And another, thing: I don't think that `.unique().order()` is possible to obtain the unique ordered categories? As `.unique` already returns a plain ndarray, and this does not know anymore about the sort order of the categorical.

**Comment by shoyer** [comment]
@jorisvandenbossche Maybe `.unique()` just should return another `Categorical` object then? It's straightforward enough to coerce to ndarray when desired via `np.asarray`.

**Comment by jorisvandenbossche** [comment]
I think we had that discussion before if I recall correctly, and the we decided to change it from categorical to ndarray ... (but I should look it up)

**Comment by jorisvandenbossche** [comment]
Update: ok, the discussion was about that it first returned an Index, and that was changed to an array.

**Comment by JanSchulz** [comment]
I don't think `.unique()` should return anything else than what the rest of pandas returns so that there are as few surprises and needs for special casing Series/cols of categorical values.

File: pandas/core/categorical.py [hunk 3]
```diff
@@ -1388,15 +1388,23 @@ def unique(self):
...
         """
-        unique_codes = np.unique(self.codes)
-        # for compatibility with normal unique, which has nan last
-        if unique_codes[0] == -1:
-            unique_codes[0:-1] = unique_codes[1:]
-            unique_codes[-1] = -1
+        from pandas.core.nanops import unique1d
+        # unlike np.unique, unique1d does not sort
+        unique_codes = unique1d(self.codes)
+        if self.ordered:
```
Comment (JanSchulz) [inline pandas/core/categorical.py]: The more I think about this (and read the bugreports which have  problems with this), the more I disagree with this change: if unique returns values in the order in which they appear in the cols, then this "contract" should be used. "order" only says that values in the categorical can be ordered, not that unique switches the order. If one wants to have the categories in order, then take the `categories` directly (e.g. https://github.com/mwaskom/seaborn/issues/361)

The only think missing is the "get me all unique values in the same order as specified by the categories", and one could give that a new keyword or a new method

**Comment by jorisvandenbossche** [comment]
@JanSchulz for that last thing you mention, you can do:

```
In [44]: s = pd.Series(pd.Categorical(list('acbab'), list('abcde')))

In [45]: s
Out[45]:
0    a
1    c
2    b
3    a
4    b
dtype: category
Categories (5, object): [a < b < c < d < e]

In [46]: s.unique()
Out[46]: array(['a', 'b', 'c'], dtype=object)  # <----- this should then be [a, c, b] 

In [49]: s.cat.remove_unused_categories().cat.categories
Out[49]: Index([u'a', u'b', u'c'], dtype='object')   # <----- this stays as it is 
```

**Comment by JanSchulz** [comment]
The last step is probably costly, as `remove_unused_categories()` does a complete redo of the factorization :-(

**Comment by jreback** [comment]
Here is a cheap way of doing the last step (ignoring that you need to take out -1 for nan and then re-add them)

```
In [25]: s.cat.categories.take(unique1d(s.cat.codes))
Out[25]: Index([u'a', u'c', u'b'], dtype='object')
```

**Comment by JanSchulz** [comment]
@jreback: nope, that's the sorting order as defined on the rows ("by apearance"), not the categories. Yours should be the implementation of normal `unique()`.

The last step above would be `s.cat.categories.take(np.unique(s.cat.codes))`, as that would be ordered in the same way as `categories`.

So basically:

```
def unique(sort=False)
   """ ...
   sort: sort the unique values by the order specified in `categories`
   """
    from pandas.core.nanops import unique1d
    # unlike np.unique, unique1d and pandas does not sort by default
    unique_codes = unique1d(self.codes)
    if sort:
        unique_codes = np.sort(unique_codes)
        # for compatibility with normal unique, which has nan last
        if unique_codes[0] == -1:
            unique_codes[0:-1] = unique_codes[1:]
            unique_codes[-1] = -1
```

**Comment by shoyer** [comment]
OK, I think we are in agreement that unique should not sort by default. I don't think that should be an option, either -- that's not found in any other unique methods, and it's better to support composability than to add flags.

Either it should return a Categorical (which is fine) or it can return an ndarray (like it currently does). In the later case, it's easy to sort by turning things back into a categorical if desired: `Categorical(c.unique(), c.categories).order()`. Given the ease of doing this and the advantages of uniform signatures, I suppose we should probably stick with ndarray.

**Comment by TomAugspurger** [comment]
+1 for no sort option and +1 for returning an ndarray.

**Comment by jorisvandenbossche** [comment]
Also +1 for no sorting at all (ordered or unordered)

For the return value: `Series.values` also returns a Categorical instead of an array, so in that regard it would be somewhat consistent within the 'categorical series' to let `Series.unique` return a Categorical, but of course not consistent within `unique`

**Comment by shoyer** [comment]
OK, I went for the simple route of making unique never sort and return an ndarray.

In principle, returning a categorical from unique is the right idea -- that would be the right decision if categorical was a real numpy dtype. But with categorical being mostly but not fully ndarray-like, it just causes more hassle for now.

PS @SteveSimmons you should setup your editor to trim trailing whitespace. It's the source of some spurious changes in this PR. Not a big deal, just something to keep in mind for next time.

File: pandas/core/categorical.py [hunk 4]
```diff
@@ -1386,17 +1386,16 @@ def unique(self):
...
         Returns
         -------
         unique values : array
         """
-        unique_codes = np.unique(self.codes)
-        # for compatibility with normal unique, which has nan last
-        if unique_codes[0] == -1:
-            unique_codes[0:-1] = unique_codes[1:]
-            unique_codes[-1] = -1
+        from pandas.core.nanops import unique1d
```
Comment (jreback) [inline pandas/core/categorical.py]: still need the nan insertion otherwise the -1 code would have meaning

Comment (shoyer) [inline pandas/core/categorical.py]: @jreback This is fed into `take_1d` which fills -1 with `fill_value`... which is happily exactly what we want here to handle NaN. That behavior is unchanged from before (and still tested). So I think this is OK?

Comment (jreback) [inline pandas/core/categorical.py]: ahh yes that's right
ok then

Comment (JanSchulz) [inline pandas/core/categorical.py]: Afaik unique sorts nan last...

Comment (shoyer) [inline pandas/core/categorical.py]: @JanSchulz Nope, unique1d does not sort NaN last. I modified the test involving NaNs to make sure.

Comment (JanSchulz) [inline pandas/core/categorical.py]: ok, `s.unique()` also not. Seems that sorting an nan handling is only done in numpy...

```
s = pd.Series([1,2,3,4,5,np.nan,6,1,2,3,4])
s.unique()
array([  1.,   2.,   3.,   4.,   5.,  nan,   6.])
```

Sorry for the noise...

File: doc/source/whatsnew/v0.16.0.txt [hunk 5]
```diff
@@ -139,6 +139,8 @@ Bug Fixes
   SQLAlchemy type  (:issue:`9083`).
 
 
+- Items in ``Categorical.unique()`` now appear in the order in which are originally found, not in sorted order (:issue:`9331`). This is consistent with the behavior for other dtypes.
+
```
Comment (JanSchulz) [inline doc/source/whatsnew/v0.16.0.txt]: Items in ``Categorical.unique()`` (and ``s.unique()`` if ``s`` is of dtype ``category``) now appear in the order in which _they_ "(?) are originally found, not in sorted order (:issue:`9331`). Formerly it was modeled after `np.unique()` which does sort the values but is now consistent with the behavior of ``unique()`` for other pandas dtypes.

Not sure about the "they"...

**Comment by shoyer** [comment]
Just realized we don't actually guarantee ordering for unique values anywhere in general. Made a new issue about that: #9346.

**Comment by shoyer** [comment]
OK, I updated the change note based on @JanSchulz's suggestion.

Any other comments before I merge this? I am inclined to stay with ndarray output for this change, though that is by no means a final decision -- we could change this to Categorical in the future if someone makes a case for that.

**Comment by jreback** [comment]
yep this is consistent with the rest of pandas so lgtm

**Comment by jorisvandenbossche** [comment]
ok for me! 
(only small side note that your formulation in the docstring does depend on the outcome of #9346)

**Comment by shoyer** [comment]
Merged.

@jorisvandenbossche agreed, may need to revisit this following #9346 (which will hopefully be before 0.16)

**Pull request merged**